dht.go

```diff
@@ -437,3 +507,49 @@ func (dht *IpfsDHT) newContextWithLocalTags(ctx context.Context, extraTags ...ta
 	) // ignoring error as it is unrelated to the actual function of this code.
 	return ctx
 }
+
+func (dht *IpfsDHT) handleProtocolChanges(ctx context.Context) {
+	// register for event bus protocol ID changes
+	ch := make(chan event.EvtPeerProtocolsUpdated, 8)
+	cancel, err := dht.host.EventBus().Subscribe(ch)
+	if err != nil {
+		return nil, err
+	}
+	defer cancel()
+
+	pmap := make(map[protocol.ID]bool)
+	for _, p := range dht.protocols {
+		pmap[p] = true
+	}
+
+	for {
+		select {
+		case e, ok := <-ch:
+			if !ok {
+				return
+			}
+			var drop, add bool
+			for _, p := range e.Added {
+				if pmap[p] {
+					add = true
+				}
+			}
+			for _, p := range e.Removed {
+				if pmap[p] {
+					drop = true
+				}
+			}
+
+			if add && drop {
+				// TODO: discuss how to handle this case
+				log.Warning("peer adding and dropping dht protocols? odd")
```

Comment:
the way this is implemented, in practice it's very, very, very unlikely that the event will contain (a) multiple protocols, and/or (b) multiple operations.

Reply:
right, but its *possible*, so what do we do when it does?

Reply:
this is tricky. the data model allows it to happen, but the implementation doesn't allow it to happen. maybe we should revisit these event structs to eliminate the * cardinality.